enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,275 @@
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+When working with bare-metal OpenShift clusters, due to the absence of storage provisioners like those in the cloud, there is a lot of manual work to be done for consuming local storage from nodes.
+The idea is to automate the manual steps that are required for consuming local storage in an OpenShift cluster by extending the Local Storage Operator (LSO).
+
+## Motivation
+
+To automatically detect local devices and provision local volumes out of them.
+
+### Goals
+
+- Automatic detection of available disks from nodes which can be used as PVs for OpenShift-cluster.
+- Respond to the attach/detach events of the disks/devices.
+- Have options for filtering particular kind of disks based on properties such as name, size, manufacturer, etc.
+
+### Non-Goals
+
+- Inherently protect against confilict with provisioners that own local devices on the same nodes automatic detection is configured to run.
+
+## Proposal
+
+The `local-storage-operator` is already capable of consuming local disks from the nodes and provisioning PVs out of them,
+but the disk/device paths needs to explicitly specified in the `LocalVolume` CR.
+The idea is to introduce a new Custom Resource, `AutoDetectVolumes`, that enables automated discovery and provisioning of local volume based PVs on a set of nodes addressed by one or more labelSelectors.
+
+### Risks and Mitigations
+
+- LSO will detect disks that contain data and/or are in-use via ensuring that the device:
+  - can be openened exclusively.
+  - is not read-only.
+  - is not removable.
+  - has no child partitions.
+  - has no FS signature.
+  - state (as outputted by `lsblk`) is not `suspended`
+- Ensuring disks aren't re-detected as new or otherwise destroyed if their device path changes.
+  - This is already ensured by the current LSO approach of consuming disks by their `UUID`
+
+## Design Details
+
+API scheme for `AutoDetectVolumes`:
+
+```go
+
+type AutoDetectVolumeList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata"`
+    Items           []AutoDetectVolume `json:"items"`
+}
+
+type AutoDetectVolume struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata"`
+    Spec              AutoDetectVolumeSpec   `json:"spec"`
+    Status            AutoDetectVolumeStatus `json:"status,omitempty"`
+}
+
+type AutoDetectVolumeSpec struct {
+    // Nodes on which the automatic detection policies must run.
+    // +optional
+    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+    // StorageClassName to use for set of matched devices
+    StorageClassName string `json:"storageClassName"`
+    // MinDeviceCount is the minumum number of devices that needs to be detected per node.
+    // If the match devices are less than the minCount specified then no PVs will be created.
+    MinDeviceCount int `json:"minDeviceCount"`
```

Comment:
Is `MinDeviceCount` not an optional field?

Reply:
is `0` equivalent to being optional?

Reply:
I don't like 0 being optional. If this field is designed to be optional, it should be called out so.

Reply:
I wasn't thinking of keeping this field as optional and yeah the default value will be `0`.

Reply:
when you say something has a "default" value if user does not supply a value, then isn't that what is called "optional"?

Reply:
I am trying to stick to how APIs are designed in k8s and we always label a field as optional if it can be skipped by the user and api-server or controller assumes a default value, if field is missing.